src/Orleans/Threading/CancellationTokenManager.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    internal static class CancellationTokenManager
+    {
+        public static void WrapCancellationTokens(object[] arguments)
+        {
+            if (arguments == null) return;
+            for (var i = 0; i < arguments.Length; i++)
+            {
+                var argument = arguments[i];
+                if (argument is CancellationToken)
+                {
+                    arguments[i] = WrapCancellationToken((CancellationToken)argument);
+                }
+            }
+        }
+
+        public static CancellationTokenWrapper WrapCancellationToken(CancellationToken ct)
+        {
+            var tokenId = Guid.NewGuid();
+            return new CancellationTokenWrapper(tokenId, ct)
+            {
+                OnSerialization = () =>
+                {
+                    RegisterTokenCallbacks(ct, tokenId);
```

Comment:
It's often difficult to judge where allocations will occur. I believe that you can avoid allocating a closure here by using `this.CancellationToken` & `this.Id`. You could alternatively accept an `Action<CancellationToken, Guid>` parameter in the constructor.

Someone please correct me if I'm wrong.